### PR TITLE
Mp4ReaderSource module - version 2

### PR DESCRIPTION
**Pull request by mohammedzakikochargi**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

**Description**
The new version of Mp4Reader has the following upgrades - 
Ordered cache of files integration - jpeg and h264
Seek refactoring - jpeg and h264
Reverse play - jpeg

**Alternative(s) considered**

Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**

Type Choose one: (Bug fix | Feature | Documentation | Testing | Other)

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/src/OrderedCacheOfFiles.cpp [hunk 1]
```diff
@@ -126,48 +126,79 @@ bool OrderedCacheOfFiles::probe(boost::filesystem::path rootDir, std::string& vi
...
 		}
+	}
 
-		videoName = dir->path().string();
+	if (mp4Files.size())
+	{
+		std::sort(mp4Files.begin(), mp4Files.end());
+	}
+
+	for(auto& mp4File : mp4Files)
```
Comment (mraduldubey) [inline base/src/OrderedCacheOfFiles.cpp:199]: videoFIle = mp4File.begin()->string();

File: base/src/OrderedCacheOfFiles.cpp [hunk 2]
```diff
@@ -126,48 +126,79 @@ bool OrderedCacheOfFiles::probe(boost::filesystem::path rootDir, std::string& vi
...
-			continue;
+			if (filePatternCheck(mp4FileIter->path()))
+			{
+				mp4Files.push_back(mp4FileIter->path());
+			}
 		}
+	}
 
-		videoName = dir->path().string();
+	if (mp4Files.size())
```
Comment (mraduldubey) [inline base/src/OrderedCacheOfFiles.cpp:194]: instead:
check size, if empty, return false
else, sort and assign first video file

File: base/src/OrderedCacheOfFiles.cpp [hunk 3]
```diff
@@ -708,13 +775,15 @@ bool OrderedCacheOfFiles::parseFiles(uint64_t start_ts, bool direction, bool inc
 			++parsedFilesCount;
 		}
 	}
+	LOG_ERROR << "video cache size = " << videoCache.size();
```
Comment (mraduldubey) [inline base/src/OrderedCacheOfFiles.cpp:778]: make it log_info

**Review by mraduldubey — CHANGES_REQUESTED** [review]

**Review by mraduldubey — CHANGES_REQUESTED** [review]
Recheck all the files & address comments

File: base/include/Command.h
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/include/Command.h:1]: yes, as we discussed we'll create a separate PR for it

File: base/include/OrderedCacheOfFiles.h
[the remark this answers is not in the document]
Comment (mohammedzakikochargi) [reply base/include/OrderedCacheOfFiles.h:1]: done

**Review by mraduldubey — APPROVED** [review]

**Pull request merged**